### PR TITLE
Intoduce endpointslice controller

### PR DESCRIPTION
**Pull request by kvaps**
Based on https://github.com/kubevirt/cloud-provider-kubevirt/pull/291
Vendor dependencies https://github.com/kubevirt/cloud-provider-kubevirt/pull/331
Fixes https://github.com/kubevirt/cloud-provider-kubevirt/issues/41

### CODE REVIEW

File: pkg/controller/kubevirteps/kubevirteps_controller.go [hunk 1]
```diff
@@ -595,7 +596,7 @@ func (c *Controller) newSlice(service *v1.Service, desiredPorts []discovery.Endp
 
 func (c *Controller) getDesiredEndpoints(service *v1.Service, tenantSlices []*discovery.EndpointSlice) []*discovery.Endpoint {
 	var desiredEndpoints []*discovery.Endpoint
-	if service.Spec.ExternalTrafficPolicy == v1.ServiceExternalTrafficPolicyTypeLocal {
+	if service.Spec.Selector == nil {
```
Comment (kvaps) [inline pkg/controller/kubevirteps/kubevirteps_controller.go:599]: This requires to support `externalTrafficPolicy: Cluster` with:

```
enableEPSController: true
selectorless: true
```

File: cmd/kubevirt-cloud-controller-manager/main.go [hunk 2]
```diff
@@ -44,14 +44,13 @@ func main() {
...
 	controllerInitializers := app.DefaultInitFuncConstructors
-	controllerAliases := map[string]string{"kubevirt-eps": kubevirteps.ControllerName.String()}
 
 	// add kubevirt-cloud-controller to the list of controllers
 	controllerInitializers[kubevirteps.ControllerName.String()] = app.ControllerInitFuncConstructor{
 		Constructor: StartKubevirtCloudControllerWrapper,
 	}
 
-	command := app.NewCloudControllerManagerCommand(ccmOptions, cloudInitializer, controllerInitializers, controllerAliases, fss, wait.NeverStop)
+	command := app.NewCloudControllerManagerCommand(ccmOptions, cloudInitializer, controllerInitializers, fss, wait.NeverStop)
```
Comment (kvaps) [inline cmd/kubevirt-cloud-controller-manager/main.go:53]: New controller manager lib have no controllerAliases

File: cmd/kubevirt-cloud-controller-manager/kubevirteps.go [hunk 3]
```diff
@@ -37,7 +39,7 @@ func startKubevirtCloudController(
 		return nil, false, err
 	}
 
-	if !*kubevirtCloud.GetCloudConfig().LoadBalancer.EnableEPSController {
+	if kubevirtCloud.GetCloudConfig().LoadBalancer.EnableEPSController == nil || !*kubevirtCloud.GetCloudConfig().LoadBalancer.EnableEPSController {
```
Comment (kvaps) [inline cmd/kubevirt-cloud-controller-manager/kubevirteps.go:42]: fixes panic if `enableEPSController` in not set

File: cmd/kubevirt-cloud-controller-manager/kubevirteps.go [hunk 4]
```diff
@@ -54,37 +56,46 @@ func startKubevirtCloudController(
...
+		if err != nil {
+			klog.Errorf("Failed to create client config from infra kubeconfig: %v", err)
+			return nil, false, err
+		}
+		restConfig, err = clientConfig.ClientConfig()
+		if err != nil {
+			klog.Errorf("Failed to create rest config for infra cluster: %v", err)
+			return nil, false, err
+		}
 	}
```
Comment (kvaps) [inline cmd/kubevirt-cloud-controller-manager/kubevirteps.go:85]: enables support for in-cluster config loading

File: pkg/controller/kubevirteps/kubevirteps_controller.go [hunk 5]
```diff
@@ -312,7 +313,7 @@ func (c *Controller) getInfraEPSFromInfraService(ctx context.Context, svc *v1.Se
 
 func (c *Controller) reconcile(ctx context.Context, r *Request) error {
 	service, ok := r.Obj.(*v1.Service)
-	if !ok {
+	if !ok || service == nil {
```
Comment (kvaps) [inline pkg/controller/kubevirteps/kubevirteps_controller.go:316]: add nil check for a service object

File: pkg/controller/kubevirteps/kubevirteps_controller.go [hunk 6]
```diff
@@ -350,8 +351,8 @@ func (c *Controller) reconcile(ctx context.Context, r *Request) error {
 	// If the services switched to a different address type, we need to delete the old ones, because it's immutable.
 	// If the services switched to a different externalTrafficPolicy, we need to delete the old ones.
 	for _, eps := range infraExistingEpSlices {
-		if service.Spec.ExternalTrafficPolicy != v1.ServiceExternalTrafficPolicyTypeLocal || serviceDeleted {
-			klog.Infof("Added for deletion EndpointSlice %s in namespace %s because it has an externalTrafficPolicy different from Local", eps.Name, eps.Namespace)
+		if service.Spec.Selector != nil || serviceDeleted {
+			klog.Infof("Added for deletion EndpointSlice %s in namespace %s because it has a selector", eps.Name, eps.Namespace)
```
Comment (kvaps) [inline pkg/controller/kubevirteps/kubevirteps_controller.go:355]: This requires to support `externalTrafficPolicy: Cluster` with:

```
enableEPSController: true
selectorless: true
```

File: pkg/provider/loadbalancer.go [hunk 7]
```diff
@@ -208,9 +208,11 @@ func (lb *loadbalancer) createLoadBalancerService(ctx context.Context, lbName st
...
 	}
 	// Give controller privilege above selectorless
-	if lb.config.EnableEPSController != nil && *lb.config.EnableEPSController && service.Spec.ExternalTrafficPolicy == corev1.ServiceExternalTrafficPolicyTypeCluster {
-		lbService.Spec.Selector = vmiLabels
-	} else if lb.config.Selectorless == nil || !*lb.config.Selectorless {
+	if lb.config.EnableEPSController != nil && *lb.config.EnableEPSController && service.Spec.ExternalTrafficPolicy == corev1.ServiceExternalTrafficPolicyTypeLocal {
+		lbService.Spec.Selector = nil
+	} else if lb.config.Selectorless != nil && *lb.config.Selectorless {
+		lbService.Spec.Selector = nil
+	} else {
```
Comment (kvaps) [inline pkg/provider/loadbalancer.go:215]: Fixes readability, and disables selector creation for `externalTrafficPolicy: Cluster` when `selectorless` mode is enabled.

**Comment by brianmcarey** [comment]
> How can I make tests passed if you need to separate PRs for vendor upgades and feature itself?

The e2e tests were failing due to a short container image name being used - the lane should be good after  https://github.com/kubevirt/cloud-provider-kubevirt/pull/332

**Comment by kvaps** [comment]
@qinqon this pr will not work without https://github.com/kubevirt/cloud-provider-kubevirt/pull/331 get merged

**Comment by qinqon** [comment]
> @qinqon this pr will not work without #331 get merged

merged
/retest

**Comment by qinqon** [comment]
@kvaps are the comments at https://github.com/kubevirt/cloud-provider-kubevirt/pull/291 addressed ?

**Comment by kvaps** [comment]
@qinqon yes

**Comment by BartVerc** [comment]
@kvaps thanks for picking this up :) 
I was once working on a refactored version, but got time issues.

**Pull request merged**